feat: add npm user validity check

maggienegm opened the pull request:
<!-- 👋 Hi, thanks for sending a PR to tidelift-me-up! 💖.
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR. -->

## PR Checklist

- [x] Addresses an existing open issue: fixes #649 
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/tidelift-me-up/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/tidelift-me-up/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

This PR adds a user validity check if the API response from the npm registry API returns zero packages to differentiate between invalid npm users and valid users with zero packages.

### How It Works

1. The npm registry API is queried with the npm username input.
2. If the response contains one or more packages, result is returned as usual. (This is the "happy" case.)
3. If the response contains 0 packages, check if npm username exists using the [npm-user wrapper](https://github.com/sindresorhus/npm-user).
4. If npm username doesn't exist, return an `Invalid npm username.` error.
6. If npm username does exist but user has zero packages, return a `No packages found for npm username: ${username}.` error.

### Testing

I updated the unit tests that were affected by these changes and then I ran the unit tests using `pnpm run test`.

🤖

JoshuaKGoldberg commented on src/tideliftMeUp.ts:
```diff
@@ -37,9 +38,22 @@ export async function tideliftMeUp({
+				throw new TideliftMeUpError(`Invalid npm username.`);
+			}
+			throw error;
+		}
```
[Refactor] I found this control flow a little confusing. When the `npmUser` call doesn't reject, it throws an error - then immediately catches it, does some checks that won't run true, and re-throws it?

Proposal: how about moving it to after the `try`/`catch`?

I.e.:

```ts
try {
  await npmUser(username);
} catch (error) {
  if ( ... ) {
    throw new TideliftMeUpError(`Invalid npm username.`);
  }
  throw error;
}

throw new TideliftMeUpError(`No packages found for npm username: ${username}.`);
```

JoshuaKGoldberg commented on src/tideliftMeUp.ts:
```diff
@@ -37,9 +38,22 @@ export async function tideliftMeUp({
+				"code" in error &&
+				error.code === "ERR_NO_NPM_USER"
+			) {
+				throw new TideliftMeUpError(`Invalid npm username.`);
```
[Feature] Could this also print the username? If I'm sending in invalid data I might not be sending the username I thought.

```suggestion
				throw new TideliftMeUpError(`Invalid npm username: ${username}.`);
```

JoshuaKGoldberg commented on src/tideliftMeUp.test.ts:
[Testing] Could you also add tests for invalid errors thrown by `npmUser`? There are a few cases in the `if` but no added tests for them.

JoshuaKGoldberg reviewed (changes requested):
A good start! I think the core idea is solid. Requesting a bit of touching up on the error, a refactor for clarity, and some testing.

maggienegm commented:
Looks like the checks are failing due to a `ERR_PNPM_BROKEN_LOCKFILE` error.

JoshuaKGoldberg commented:
Weird, looks like it was some kind of merge weirdness from #669. I fixed it on `main`.

maggienegm commented:
Thank you! I updated this branch with `main`, and during the merge process, the version number of one of the dependencies didn't update for some reason (it didn't show in the merge conflicts?). But I fixed it and now this PR should be good to go!

JoshuaKGoldberg commented on src/tideliftMeUp.ts:
```diff
@@ -61,6 +69,25 @@ export async function tideliftMeUp({
 
+async function doesUserExist(
+	username: string,
+): Promise<"invalid username" | "user not found" | "user with no packages"> {
```
💡 Sorry, I had one more thought on how to make this flow more clear. Right now there are two places in code that each handle the three possible errors. This `doesUserExist` function has three different `return`s, and then there are three different `throw new Error`s up in `tideliftMeUp`. That's kind of duplicated logic, no?

Proposal: how about making this function the one that creates an error, and renaming it to something like `createPackagesError` / `createUserError`?

```ts
function createPackagesError(username: string): Promise<Error>
```

This way the code above can just call it to get an error, without having to handle three separate cases. And then we don't have this ad hoc string literal hanging around.

WDYT?

maggienegm replied:
I agree, this is a much cleaner and clearer approach 👍

JoshuaKGoldberg reviewed (changes requested):
Looking great, just one more refactor to talk about!

JoshuaKGoldberg reviewed (approved):
This is probably the first time I've seen `return throw await` actually used 😄 very nice! 🔥
